akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -541,8 +708,9 @@ private[stream] object MaterializerSession {
 private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Module, val initialAttributes: Attributes) {
   import StreamLayout._
 
-  private var subscribersStack: List[ju.Map[InPort, Subscriber[Any]]] =
-    new ju.HashMap[InPort, Subscriber[Any]] :: Nil
+  // the contained maps store either Subscriber[Any] or VirtualPublisher, but the type system cannot express that
```

Comment:
I debated that, but the lack of type-safety was what led me down this alternative path: now it is painfully obvious that this map does not only contain suitable Subscribers but also some weird things.